api/layer/multipart_upload.go

```diff
@@ -1184,3 +1250,213 @@ func (n *layer) manualSlice(ctx context.Context, bktInfo *data.BucketInfo, prm P
 
 	return id, elements, nil
 }
+
+// uploadPartAsSlot uploads multipart part, but without correct link to previous part because we don't have it.
+// It uses zero part as pivot. Actual link will be set on CompleteMultipart.
+func (n *layer) uploadPartAsSlot(ctx context.Context, params uploadPartAsSlotParams) (*data.ObjectInfo, error) {
+	zeroPart, err := n.treeService.GetPartByNumber(ctx, params.bktInfo, params.multipartInfo.ID, 0)
+	if err != nil {
+		return nil, fmt.Errorf("get part by number: %w", err)
+	}
+
+	var (
+		id              oid.ID
+		chunk           *[]byte
+		elements        []data.LinkObjectPayload
+		isReturnToPool  bool
+		splitFirstID    = zeroPart.OID
+		splitPreviousID = zeroPart.OID
+		multipartHash   = sha256.New()
+		currentPartHash = sha256.New()
+	)
+
+	objHashes := []hash.Hash{multipartHash, currentPartHash}
+	if params.tzHash != nil {
+		objHashes = append(objHashes, params.tzHash)
+	}
+
+	params.attributes = append(params.attributes,
+		[2]string{headerS3MultipartUpload, params.multipartInfo.UploadID},
+		[2]string{headerS3MultipartNumber, strconv.FormatInt(int64(params.uploadPartParams.PartNumber), 10)},
+		[2]string{headerS3MultipartCreated, strconv.FormatInt(time.Now().UnixNano(), 10)},
+	)
+
+	prm := PrmObjectCreate{
+		Container:    params.bktInfo.CID,
+		Creator:      params.bktInfo.Owner,
+		Attributes:   params.attributes,
+		CreationTime: params.creationTime,
+		CopiesNumber: params.multipartInfo.CopiesNumber,
+		Multipart: &Multipart{
+			MultipartHashes: objHashes,
+		},
+	}
+
+	if params.uploadPartParams.Size > n.neoFS.MaxObjectSize()/2 {
+		chunk = n.buffers.Get().(*[]byte)
+		isReturnToPool = true
+	} else {
+		smallChunk := make([]byte, params.uploadPartParams.Size)
+		chunk = &smallChunk
+	}
+
+	id, elements, err = n.manualSlice(ctx, params.bktInfo, prm, splitFirstID, splitPreviousID, *chunk, params.payloadReader)
+	if isReturnToPool {
+		n.buffers.Put(chunk)
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("manual slice: %w", err)
+	}
+
+	partInfo := &data.PartInfo{
+		Key:      params.uploadPartParams.Info.Key,
+		UploadID: params.uploadPartParams.Info.UploadID,
+		Number:   params.uploadPartParams.PartNumber,
+		OID:      id,
+		Size:     params.decSize,
+		ETag:     hex.EncodeToString(currentPartHash.Sum(nil)),
+		Created:  prm.CreationTime,
+		Elements: elements,
+	}
+
+	// encoding hash.Hash state to save it in tree service.
```

Comment:
I think it's not necessary for these parts. These hashes can't be reused.

Reply:
this method works for all types of parts, for complete reupload too. I prefer to have the single method without extra `if`